plotting|cmds: Fix and improve `chia plots check`

xdustinface opened the pull request:
Right now it has an issue which leads to false negatives imo. As it 
currently breaks the inner testing loop as soon as one challange did not 
result in valid or even in a proof at all. But to determine if a plot is 
valid i think it should try all provided challenges first and then 
decide if the plot is valid by checking if there were any proofs 
available for the given challenges, no?

So this commit removes `caught_exception` and it's check/break if an 
exception occurs. Instead it just prints the error as debug message and 
it increments `total_proofs` only in case of success.

xdustinface committed:
plotting: Fix check_plots.py

Right now it has an issue which leads to false negatives. As it 
currently breaks the inner testing loop as soon as one challange did not 
result in valid or even in a proof at all. But to determine if a plot is 
valid i think it should try all provided challenges first and then 
decide if the plot is valid by checking if there were any proofs 
available for the given challenges, no?

So this commit removes caught_exception and its check/break if an 
exception occurs. Instead it just prints the error as debug message and 
it increases total_proofs in case of success.

mariano54 commented:
@eFishCent can you take a look at this? Do you remember why you added a break, if an exception was thrown in the prover?

eFishCent commented:
So the assertion was put in to catch bit flips.
The exception catches truncated files, usually copies that didn't finish, but passes the bit flip tests.
I have test plots specifically created to verify `chia plots check -n 200` is working properly shared on dropbox: https://www.dropbox.com/sh/y7vzfz1nx5iy6w1/AAC5_FBFFHuNj_FWbGSRaiRXa?dl=0
Please download and test any changes to make sure both conditions are caught in any changes to `chia plots check`

Thanks!
@eFishCent

hoffmang9 commented:
We should pull those tests into the test harness. We can check your test plots into the test-cache repo using a similar flow to the test plots.

eFishCent commented:
I should note, bitflips are probably ok for the prover, but the truncates (and other unknown problems) do cause prover to crash. This can potentially take down harvesters, which is why I wanted to make sure `chia plots check` flagged those.

I recommend if you are to modify `chia plots check` to be forgiving, only do it for bitflips.

eFishCent commented:
> Right now it has an issue which leads to false negatives imo. As it
> currently breaks the inner testing loop as soon as one challange did not
> result in valid or even in a proof at all. But to determine if a plot is
> valid i think it should try all provided challenges first and then
> decide if the plot is valid by checking if there were any proofs
> available for the given challenges, no?
> 
> So this commit removes `caught_exception` and it's check/break if an
> exception occurs. Instead it just prints the error as debug message and
> it increments `total_proofs` only in case of success.

One last thought. If you are to be forgiving for bit flips, please make sure there is a command line switch to allow `--strict` mode, so for those who want 100% accuracy on `chia plots check` still have access to it.

xdustinface committed:
plotting|cmds: Introduce -s/--success-rate for `chia plots check`

xdustinface committed:
cmds: Add short params for --debug-show-memo and --challenge-start

xdustinface committed:
plotting: Improve exception handling

- Directly catch KeyboardInterrupt
- Catch Exception in other cases and print the error

xdustinface committed:
plotting: Improve error message if quality doesn't match

xdustinface committed:
plotting: Print the progress based on number of already checked plots

xdustinface committed:
plotting: Introduce and use min_num and default_num

xdustinface committed:
plotting|cmds: Introduce -r/--random as option to use random challenges

xdustinface commented:
> We should pull those tests into the test harness. We can check your test plots into the test-cache repo using a similar flow to the test plots.

Sounds like a good plan! Do it! :D 

> One last thought. If you are to be forgiving for bit flips, please make sure there is a command line switch to allow `--strict` mode, so for those who want 

See fd59c7ad0f2ff63c100e1ff285cb924d3189184f, which adds `-s/--success-rate` to control how accurate you want it to be (default is 50% which turned out to be a good value IMO with n=30).

> 100% accuracy on `chia plots check -n [challenges]` still have access to it.

In general, from my understanding its not a good idea to expect 100% accuracy at all because not all plots have a proof for all challenges if i got it right. It might happen (and it does seem to happen quite often) that you end up with 0 valid proofs for a challenge even though the plot is valid. 


> do cause prover to crash. This can potentially take down harvesters

Sounds like the error handling in the harvester rather needs to be improved then :) IMO it should really not rely on users putting only valid plots in their directories. Can you link to where you think it could crash? Also, do you have/can you generate other plots which tigger the other cases you are talking about?


Note: The last push also added some more commits with other improvements.

hoffmang9 reviewed (approved):
Looks good to me

eFishCent commented:
> In general, from my understanding its not a good idea to expect 100% accuracy at all because not all plots have a proof for all challenges if i got it right. It might happen (and it does seem to happen quite often) that you end up with 0 valid proofs for a challenge even though the plot is valid.

I want to clarify a confusion point, in  where we make assert and trap exceptions means there is an actual error in the plot file. If the blockchain actually requested that specific challenge, the lookup would fail for that plot, where it should have potentially succeeded in getting a proof.

We defaulted `-n 30` as it is statistically almost impossible to not find at least one proof with 30 challenges.n @mariano54 can correct me if I am wrong about this.

This PR actually creates a sense that a plot file is healthy when it isn't. @hoffmang9 and @mariano54 Are you sure we want to do this?

eFishCent commented:
> Sounds like the error handling in the harvester rather needs to be improved then :) IMO it should really not rely on users putting only valid plots in their directories. Can you link to where you think it could crash? Also, do you have/can you generate other plots which tigger the other cases you are talking about?

This is something that @mariano54 is aware of but we haven't been able to prevent it from happening as we don't have plots that can reproduce this problem. We tried to get real world bad plots saved at some point, but I left that up to @justinengland to collect from people who reported having bad plots. Just to be clear, bad plots almost 100% of the time occur from bad hardware or overclocking that cause bit flips in the wrong places that cause harvester to crash. Some bit flips are recoverable by the harvester, but some aren't. My creation of the truncated file was to simulate the unexpected bit flips in large amounts to cause the 2nd type of exceptions.

mariano54 commented:
> We defaulted -n 30 as it is statistically almost impossible to not find at least one proof with 30 challenges with low difficulty. When the blockchain is running, the difficulty will always be higher than what chia plots check is set at, which is why not all plots will have proofs for a challenge. @mariano54 can correct me if I am wrong about this.

This is not exactly right: There is always an expected value of 1 proof for any challenge. In the harvester, we don't actually submit the proof to the chain though, unless it qualifies to be a winner (so 99.999% of proofs are just discarded and never logged).

I agree with @eFishCent that it's about finding bad plots. However, I also agree with @xdustinface that if we have enough plots returning valid proofs, we can consider this plot as valid. Maybe 15/30 is not enough though, I have seen people getting 7/30 etc. Maybe we should change to 25/50 for more statistical validity.

eFishCent commented:
> > We defaulted -n 30 as it is statistically almost impossible to not find at least one proof with 30 challenges with low difficulty. When the blockchain is running, the difficulty will always be higher than what chia plots check is set at, which is why not all plots will have proofs for a challenge. @mariano54 can correct me if I am wrong about this.
> 
> This is not exactly right: There is always an expected value of 1 proof for any challenge. In the harvester, we don't actually submit the proof to the chain though, unless it qualifies to be a winner (so 99.999% of proofs are just discarded and never logged).

Thanks for correcting my misunderstanding!

> I agree with @eFishCent that it's about finding bad plots. However, I also agree with @xdustinface that if we have enough plots returning valid proofs, we can consider this plot as valid. Maybe 15/30 is not enough though, I have seen people getting 7/30 etc. Maybe we should change to 25/50 for more statistical validity.

I agree with one caveat: if the original exception code I wrote (for finding bit flips and truncated files) catch any errors, we flag that plot immediately as invalid. Both these errors will at a minimum reduce the chances for the plot to farm XCH, at worst unintended behavior in the harvester may happen.

hoffmang9 commented:
Keep alive.

mariano54 commented:
Can we close @xdustinface ?

xdustinface commented:
Hm well yeah you can close it. I think there were a lot changes related to this in the meantime. I will put it on my todo list to review this again at some point!
